### PR TITLE
fix(runtime): replace correlation-id-space routing with a structured request path

### LINKED ISSUE #4808: Per-route filter's out-of-band sendRequest() response is delivered to the client, not the filter, when 2+ routes send OOB requests on one connection
**Describe the bug** 

With **dynamic routing**, a per-route filter that issues an out-of-band (OOB) request via `FilterContext.sendRequest(...)` never gets its response when **2+ routes issue an OOB request on the same client connection**. The broker's response comes back correctly but is delivered to the **downstream client** instead of completing the filter's promise, which times out after 20s.

**Cause:** all OOB requests share one correlation id (`CorrelationIdSpace.RESERVED_OUT_OF_BAND_CORRELATION_ID = Integer.MIN_VALUE`), and `RoutingTerminalHandler` keys its `correlationId -> routeName` map on it. With 2+ concurrent OOB
requests the entries collide, so the reply is tagged with the wrong route; `RouteFilterHandler.write` only completes an OOB reply when the frame's `routeName` matches its route, so it's never handed back. (A single OOB request per connection works fine.)

Reproduced on `0.24.0-SNAPSHOT`, commit `50919886` (= current `main` HEAD).

---

**To Reproduce**

1. Virtual cluster with **dynamic routing** and **two routes** to different brokers.
2. A **per-route filter** on each route whose `onRequest` calls `context.sendRequest(...)` (any request expecting a response) and awaits it.
3. Client sends a request the router fans to **both** routes (e.g. `Metadata`) -> both filters issue their OOB request on the same connection at once.
4. Both time out: `Asynchronous ... request made by filter '...[route=...]' failed to complete within 20000 ms`.

---

**Expected behavior**

Each OOB response is delivered back to the filter that issued it (promise completes), regardless of how many routes issue OOB requests concurrently. The response frame already carries its `recipient` filter.

---

**YAML files and logs**

One connection, two OOB legs. Both send OOB with `correlationId=-2147483648`; both brokers answer `errorCode=0`; but leg a's reply (recipient = filter a) is written to the **client** channel:

```
Upstream  [-> broker A] WRITE InternalRequestFrame(SASL_HANDSHAKE, correlationId=-2147483648)
Upstream  [<- broker A] READ  InternalResponseFrame(recipient=...Filter@712d0d1c, promise=[Not completed], errorCode=0)
Downstream[-> client]   WRITE InternalResponseFrame(recipient=...Filter@712d0d1c, promise=[Not completed], errorCode=0)   <-- wrong destination
... 20s later: Asynchronous SASL_HANDSHAKE request ... failed to complete within 20000 ms
```
---

**Additional context**

One client connection = one shared pipeline; both per-route filters and one `RoutingTerminalHandler` (with the shared map) live on it:

```
   +---------------------------------------------------------------+
   |  RouteFilterHandler[route=a]  (per-route filter a)            |
   |  RouteFilterHandler[route=b]  (per-route filter b)            |
   |  RoutingTerminalHandler   map: correlationId -> routeName     |
   +---------------------------------------------------------------+
          |  OOB (id=MIN)                 |  OOB (id=MIN)
          v                               v
      [ broker A ]                    [ broker B ]
          |                               |
   reply(id=MIN, recipient=a)     reply(id=MIN, recipient=b)
          |                               |
          +--------------> written to the CLIENT  <-- BUG (should go to filter a / b)
```

Why (both OOB requests use the same map key `MIN`):

```
1. filter a sendRequest -> map.put(MIN,"a")
2. filter b sendRequest -> map.put(MIN,"b")     map = { MIN:"b" }   <-- COLLISION
3. broker A reply       -> map.remove(MIN)="b"  (reply tagged WRONG route)
4. RouteFilterHandler[a]: "b"!="a" -> pass through, not completed
5. no handler completes it -> written to the CLIENT
6. filter a's promise never completes -> 20s timeout   (filter b: same)
```
---

Code (module `kroxylicious-runtime`):
- [`RouteFilterHandler.write`](https://github.com/kroxylicious/kroxylicious/blob/50919886b75348a50cea9b8e5ff9a08a397b18e8/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/RouteFilterHandler.java#L57-L64) gates OOB completion on [`matchesRoute`](https://github.com/kroxylicious/kroxylicious/blob/50919886b75348a50cea9b8e5ff9a08a397b18e8/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/RouteFilterHandler.java#L66-L68)
- `RoutingTerminalHandler` `correlationId -> routeName` map: [decl](https://github.com/kroxylicious/kroxylicious/blob/50919886b75348a50cea9b8e5ff9a08a397b18e8/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingTerminalHandler.java#L44), [put](https://github.com/kroxylicious/kroxylicious/blob/50919886b75348a50cea9b8e5ff9a08a397b18e8/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingTerminalHandler.java#L66), [remove](https://github.com/kroxylicious/kroxylicious/blob/50919886b75348a50cea9b8e5ff9a08a397b18e8/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingTerminalHandler.java#L96)
- [`CorrelationIdSpace.RESERVED_OUT_OF_BAND_CORRELATION_ID`](https://github.com/kroxylicious/kroxylicious/blob/50919886b75348a50cea9b8e5ff9a08a397b18e8/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/CorrelationIdSpace.java#L15) (single `Integer.MIN_VALUE`)
- completion path: [`handleInternalResponseWrite` -> `isRecipient`](https://github.com/kroxylicious/kroxylicious/blob/50919886b75348a50cea9b8e5ff9a08a397b18e8/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java#L156-L160) -> [`completeInternalResponse`](https://github.com/kroxylicious/kroxylicious/blob/50919886b75348a50cea9b8e5ff9a08a397b18e8/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java#L625)

---

**Suggested fixes**

- **Deliver OOB responses by recipient, not by route (minimal).** In `RouteFilterHandler.write`, let an `InternalResponseFrame` bypass the `matchesRoute` gate so `super.write` (`isRecipient`) can complete it (it's self-addressed, so `routeName` is irrelevant):
  ```java
  if (msg instanceof InternalResponseFrame || matchesRoute(msg)) {
      super.write(ctx, msg, promise);
  } else {
      ctx.write(msg, promise);
  }
  ```

- **Give OOB requests distinct correlation ids (complementary hardening).** Allocate OOB ids from a reserved range (as already done for routing ids) instead of the single `Integer.MIN_VALUE` constant, so the `correlationId -> routeName` map can't collide.

Happy to open a PR.

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Bugfix

### Description

Get rid of the whole idea of "reserved correlation ids ranges" previously used for tracking OOB requests. The sheer complexity of that mechanism was an underlying cause for #4808. Instead build on the existing machinery we have for tracking the recipients for internal `Frames` via the `CorrelationManager` and `Correlation` classes. Rather than recording the name of the recipient on the `Frame` we now build a path (`PathElement`) as a request (whether originating from the client or a filter or router) traverses the router graph. This path is attached to the request `Frame` and passed via the `Correlation` to the response `Frame`. Because the path has captured how the request was routed and the originator we can use than when dispatching the response. For OOB responses we know the filter or router which was the sender of the original request, so we know to stop forwarding such responses back to the client end of the Netty pipeline.

In addition to fixed #4808 this also addresses at least three other issues:

1. Routes with concurrent OOB requests in flight suffered from a similar collision to that which caused #4808
2. The correlation id was exposed to filters and routers, so using the single sentinel as we were previously prevented those plugins from correctly correlating OOB requests and responses, 
3. We were previously depending on the equality of Filters, thus assuming the Filters did not override `equals()`.

Fixes #4808.

### CODE REVIEW

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
```
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java:13]: Could we add some unit tests for PathElement?  This will also help aid understand of how it works.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java [hunk 2]
```diff
@@ -58,17 +58,17 @@ public interface Frame {
     boolean isDecoded();
 
     /**
-     * The route this frame is associated with.
-     * @return The route this frame is associated with, or {@code null} if not yet routed.
+     * The position in the routing/filter tree this frame is associated with.
+     * @return The path this frame is associated with, or {@code null} if not yet routed.
      */
-    default @Nullable String routeName() {
+    default @Nullable PathElement path() {
```
Comment (hrishabhg) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java]: (optional) 
```suggestion
    default @Nullable PathElement containingPathElement() { // or container
```

**Comment by robobario** [comment]
at a guess your CI isn't running because the branch can't merge into main cleanly

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+ * A position in the routing/filter tree, from the top-level virtual cluster down to wherever a
+ * frame currently sits (or, for an internally-issued request, down to the specific filter or
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java]: I find the naming of the methods jarring given we talk about the path "_from the top-level virtual cluster down to wherever a frame currently sits_", and then `next()` runs in the opposite direction. Maybe we could call the method `parent()` to make the tree nature more solid, we already use the ancestor naming.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingHandler.java [hunk 4]
```diff
@@ -704,28 +715,17 @@ private void notifyRequestComplete() {
     @SuppressWarnings("FutureReturnValueIgnored")
     @Override
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        // No defensive "unhandled response with a routing-range id" check is needed any more:
+        // a router-issued request's promise is carried directly on its own frame's path, reachable
+        // only through that frame, so there's no separate id-keyed lookup step left to miss. And no
+        // per-hop route retagging is needed either - the path was already correct (full lineage
+        // included) when it was set once, at dispatch time.
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingHandler.java]: will relay this from Claude as it put this forward as a priority, I haven't looked into it

> 1. Removed defensive close is a real safety-net loss (RoutingHandler.write, ~715). Old code: an UNHANDLED response with an in-range routing correlation id closed the connection. Now removed. Argument ("promise reachable only through the frame's path") is sound for the happy path, but the net previously caught bugs elsewhere. Now a stray/duplicate Router-leaf frame that fails handleResponse's parent-match falls through to ctx.write(msg) — written outbound toward the client instead of closing. Worst case a raw internal frame leaks to the client rather than a clean disconnect. Low probability, higher blast radius. Consider keeping a fail-closed guard: if a frame still carries a PathElement.Router/Filter leaf after all levels declined it, close rather than forward.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java [hunk 5]
```diff
@@ -155,14 +157,56 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
...
+     * dynamic dispatch grafts a deeper route onto the same leaf - see
+     * {@code RoutingHandler.dispatchStaticRoute}). A VC-level (non-route-scoped) handler has
+     * exactly one instance for the whole connection, so route position carries no identity for it
+     * at all - it matches on name and ordinal alone.
+     */
+    boolean isRecipient(Frame msg) {
+        return msg.path() instanceof PathElement.Filter f
+                && f.name().equals(filterAndInvoker.filterName())
+                && f.ordinal() == ownOrdinal()
+                && (ownRoutePath() == null || ownRoutePath().isAncestorOfOrSameAs(f.next()));
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: It feels like we should be able to say ` msg.path().equals(ownPath())`. If every FilterHandler has a correct path.

I guess in that case FilterHandler might not need ownOrdinal as it should be part of its path?

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java [hunk 6]
```diff
@@ -155,14 +157,56 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
...
+    }
+
+    /**
+     * This handler's own route position, or {@code null} if it is not route-scoped. Subclasses
+     * (see {@code RouteFilterHandler}) override this to identify out-of-band requests/responses
+     * as belonging to a specific route.
+     */
+    @Nullable
+    PathElement ownRoutePath() {
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: as above, for uniformity maybe all FilterHandlers should return a correct `ownPath` even though they might not be needed on the non-routed case.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java [hunk 7]
```diff
@@ -155,14 +157,56 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
...
+        return null;
+    }
+
+    /**
+     * This handler's filter's position within its enclosing filter list, disambiguating two
+     * filters that happen to share a configured name. Defaults to {@code 0}; overridden by
+     * route-scoped handlers.
+     */
+    int ownOrdinal() {
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: I wonder if every filterHandler has a correct ownPath, can we delete this? it would be encoded into the path.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+ */
+public sealed interface PathElement {
+
+    /**
+     * The next element toward the top-level virtual cluster, or {@code null} if this is the
+     * outermost element.
+     *
+     * @return the next element, or {@code null}
+     */
+    @Nullable
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java]: I think a null object pattern with a VirtualClusterRoot PathElement might make this more comprehensible in future. Unrouted frames could have a VirtualClusterRoot originator and the nulls won't spider out across the land. At the cost of needing a self-reference at the root.

It doesn't buy much/anything functionally but I'd rather see this VC root concept explicitly in code for understanding/reading reasons.

**Review by robobario — COMMENTED** [review]
Thanks for putting this together Tom, the correlation ID space management and generally relying on int identifiers looks like a dead end in retrospect.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java (hunk 3 above)
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java]: annnnd it's fixed

Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java]: Just pushed a refactoring which changed a lot of the (awful) names.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java [hunk 9]
```diff
@@ -155,14 +157,56 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
...
+     * static or dynamic dispatch grafts a deeper route onto the same leaf - see
+     * {@code RoutingHandler.dispatchStaticRoute}). A VC-level (non-route-scoped) handler has
+     * exactly one instance for the whole connection, so route position carries no identity for it
+     * at all - it matches on name and ordinal alone.
+     */
+    boolean isRecipient(Frame msg) {
+        return msg.path() instanceof PathElement.FilterOrigin f
+                && f.name().equals(filterAndInvoker.filterName())
+                && f.ordinal() == ownOrdinal()
+                && (ownRoutePath() == null || ownRoutePath().isAncestorOfOrSameAs(f.parent()));
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: why the ancestor check? 

> Determines whether {@code msg} is an internal (out-of-band) response addressed to this handler's own filter.

shouldn't the frame be carrying this FilterHandler's path if it originated here? or am I misreading the intention of this method.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java (hunk 2 above)
Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java]: Thanks for taking a look @hrishabhg, `PathElement` got refactored so things had more sensible names.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java (hunk 5 above)
Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: A `FilterOrigin`'s `parent()` can be extended after the request is issued. When the OOB request targets a route that resolves through a nested router, `RoutingHandler.dispatchStaticRoute()`/`graft()` grafts a deeper route onto the same leaf before it's dispatched further. That means the path the request left with is an ancestor of the path it comes back on, not an exact match. That's why `isRecipient` deliberately uses `ownRoutePath().isAncestorOfOrSameAs(f.parent())` rather than equality.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java (hunk 7 above)
Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: `ownOrdinal()`/`ownRoutePath()` can't be deleted because the ordinal and route are a part of the identity a FilterHandler needs (disambiguating same-named filters in a list, and disambiguating same-named/ordinal filters on different routes).

Claude reckons:
> What could change is presentation: instead of two separately-overridable methods,
  RouteFilterHandler could hand its ordinal+route to the base constructor as fields,
  dropping the override pattern — but that's a cosmetic tidy, not a behavioural
  simplification

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingHandler.java (hunk 4 above)
Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingHandler.java]: Will push a test

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java (hunk 9 above)
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java]: answered elsewhere that the frame path can accumulate more information

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java (hunk 8 above)
Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java]: Done

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingHandler.java (hunk 4 above)
Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RoutingHandler.java]: Done

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/CorrelationIdAllocator.java
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/CorrelationIdAllocator.java:1]: `inRange` of this class now has no production usages

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java (hunk 2 above)
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java]: The javadoc/name isn't very self-explanatory. It's a path, but to what? `this frame is associated with` what does it mean for a frame to be associated with a path.

Is it something like the `lastOriginatorPath` where an originator is a _thing_ that frames can originate from. Or `frameSenderPath` the path to the element that sent this frame?

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+ */
+public sealed interface PathElement {
+
+    /**
+     * The route this element sits on, or the next element toward the top-level virtual cluster,
+     * or {@link ClientOrigin} if this is the outermost element.
+     *
+     * @return the parent position
+     */
+    RoutePosition parent();
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java]: feels a bit odd that we have PathElement where only a subset of types can be parents, so we have *Origin types that can hold a promise, and then RoutePosition types that are purely positional. Just feels a bit odd. The Path encodes this head-with-promise that gets grafted about, with a parent which is more like the positional path that the frame has navigated through the routing tree.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/Configuration.java [hunk 11]
```diff
@@ -130,8 +130,12 @@ private static void validateRouterDefinitions(@Nullable List<RouterDefinition> r
...
             for (var route : router.routes()) {
                 if (route.filters() != null) {
-                    checkNamedFiltersAreDefined(filterDefsByName, route.filters(),
-                            "routerDefinitions." + router.name() + ".routes." + route.name() + ".filters");
+                    String componentName = "routerDefinitions." + router.name() + ".routes." + route.name() + ".filters";
+                    checkNamedFiltersAreDefined(filterDefsByName, route.filters(), componentName);
+                    // A duplicate filter name within one route's own list would be structurally
+                    // ambiguous for out-of-band request delivery, which disambiguates same-named
+                    // filters on a route by their position.
+                    validateUniqueNames(route.filters(), Function.identity(), componentName);
```
Comment (robobario) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/Configuration.java]: I don't follow this. The comment says it disambiguates by position so why do we need unique names?

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java (hunk 2 above)
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java]: meshes a bit with my other questions, what is it? There's a positional aspect, tracking the Routing path a frame traversed, and then it carries the promise from the origin. So is this thing really a `Path`? Or is it an `Originator` or `FrameOrigin` that has a `Path`? Something like that.

**Review by robobario — COMMENTED** [review]
would be good to update the PR description to describe how the mechanism works, what we are trying to track on a frame, currently you have to read quite a lot of piece it back together.

**Comment by robobario** [comment]
PS I'm also generally happy that this is a step forward with new test coverage in place so I don't think we need to get the typing perfect, so if you find a good point to say it's good enough Tom that's okay by me. But I would like to make sure some of the names/javadocs explain what these concepts mean before merge.

**Comment by robobario** [comment]
put up [a PR](https://github.com/tombentley/kproxy/pull/71) adding some more scenarios to the IT

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/Configuration.java (hunk 11 above)
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/Configuration.java]: This is my only remaining open question. If we need this restriction we should cover it with tests.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java (hunk 1 above)
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java:13]: this now has unit test coverage

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java (hunk 2 above)
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/Frame.java]: big docs update and type refactoring has been pushed up, this is much improved now

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java (hunk 10 above)
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/frame/PathElement.java]: refactored heavily, thread is done

**Review by robobario — APPROVED** [review]
Thanks Tom, only outstanding question is about the configuration change. If we add that we should test it somewhere.

I think escaping the meaningful correlation id ranges is a great improvement, I had worries in the past about other edge cases like layered proxies, misbehaving clients or long running kafka clients wrapping their correlations into our 'special ranges'. This removes that class of problems. 

Though we still have some issues with Filters being pushed into using correlationId to .. correlate, bugs could creep in due to our established `-1` correlationIds for OOB requests. I might have a think about whether we can update the correlated data API proposal, maybe this Path would be a good place to stash correlated request data that needs to be available to exactly the corresponding response.

**Comment by tombentley** [comment]
@robobario I think I'd prefer to get rid of that restriction, to be honest. Each difference in the rules/behaviour of filters on the VC and filters on routes just makes things conceptually more complicated. I'll see what we can do to remove it.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/CorrelationIdAllocator.java [hunk 12]
```diff
@@ -30,7 +29,7 @@ public class CorrelationIdAllocator {
      * @param maxExc upper bound, exclusive
      * @throws IllegalArgumentException if {@code minInc >= maxExc}
      */
-    CorrelationIdAllocator(int minInc, int maxExc) {
+    public CorrelationIdAllocator(int minInc, int maxExc) {
```
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/CorrelationIdAllocator.java:33]: We no longer need the three arg CorrelationIdAllocator ctor.

Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/CorrelationIdAllocator.java:33]: Fixed

**Comment by k-wall** [comment]
(Discovered with the help of Claude) Reviewing this branch, I think there's a correctness bug in the nested-router out-of-band (OOB) path.

**Scenario:** a router issues an OOB request (`RouterContext.sendToAnyNode`) to a route that itself targets a **nested** router `N`.

1. The issuing dispatcher stamps `RouterOriginator(promise, Route("routeR", ClientOrigin))`.
2. `N` intercepts and statically routes onward to its own `innerRoute`; `dispatchStaticRoute` → `graft` deepens the originator's position to `Route("N/innerRoute", Route("routeR", ClientOrigin))`.
3. On the response, `RouteDispatcher.handleResponse` matches purely on the grafted position (`routeSegment.parent() == parentPath`), so the match now fires at **N**, not the issuing level. `N` completes the outer router's promise — translating with **N's** `NodeIdMapping`/route rather than the issuer's — and the issuing dispatcher never removes the future from its own `pendingPromises` (leaks until connection close).

This contradicts the invariant your own comment in `RoutingHandler.write` asserts: a `RouterOriginator` is *"only ever consumed by the dispatcher level that issued it."*

I've put a reproducing test + candidate fix up against this branch: tombentley/kproxy#72. The test (`nestedDispatcherMustNotConsumeAnOuterRouterOriginator`) is red before the fix and green after.

The fix mirrors how `FilterOriginator` already survives grafting via its stable `name`/`ordinal`: give `RouterOriginator` a stable `issuedAt` route that `reposition` preserves, and match/translate on `issuedAt` instead of the graft-mutated position. Happy to adjust if you'd prefer a different approach (e.g. not grafting `RouterOriginator` positions at all).

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RouteDispatcher.java [hunk 13]
```diff
@@ -184,104 +219,97 @@ private CompletableFuture<ApiMessage> doSendToAnyNode(String route, RequestHeade
...
     ResponseOutcome handleResponse(DecodedResponseFrame<?> frame, String sessionId) {
-        int correlationId = frame.correlationId();
-        PendingResponse pending = pendingResponses.remove(correlationId);
-        if (pending != null) {
+        PathElement routing = frame.routing();
+        if (routing instanceof PathElement.RouterOriginator(CompletableFuture<?> promise, PathElement.Route routeSegment)
+                && Objects.equals(routeSegment.parent(), parentPath)) {
+            @SuppressWarnings("unchecked")
+            CompletableFuture<ApiMessage> future = (CompletableFuture<ApiMessage>) promise;
+            pendingPromises.remove(future);
```
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/routing/RouteDispatcher.java]: Let's check the return value of remove.  It should always be true.  It is is false, something is bust.  Let's fail early.  It'll be easier to diagnose that rather than an inappropriate ResponseOutcome.CONSUMED.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/CorrelationIdAllocator.java (hunk 12 above)
Comment (k-wall) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/CorrelationIdAllocator.java:33]: If I read the code correctly, we no longer need the case where minInc != initial, so the we may as well remove the tests and the 3 arg ctor. It is dead code.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/RouteFilterCorrectnessIT.java [hunk 14]
```diff
@@ -221,6 +236,345 @@ void routeFilterSendRequestWorksWithSingleRoute() {
...
+                        "name", markerB,
+                        "forwardingStyle", ForwardingStyle.ASYNCHRONOUS_REQUEST_TO_BROKER)
+                .build();
+
+        try (var tester = KroxyliciousTesters.mockKafkaKroxyliciousTester(
+                mockBootstrap -> singleRouteConfig(mockBootstrap, List.of(markerADef, markerBDef)), ROUTING_ENABLED);
+                var client = tester.simpleTestClient()) {
+
+            ApiVersionsResponseData apiVersions = new ApiVersionsResponseData();
+            apiVersions.apiKeys().add(new ApiVersionsResponseData.ApiVersion()
```
Comment (k-wall) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/RouteFilterCorrectnessIT.java]: Fetch mocking is redundant.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/RouteFilterCorrectnessIT.java [hunk 15]
```diff
@@ -221,6 +236,345 @@ void routeFilterSendRequestWorksWithSingleRoute() {
...
+                        "name", filterName,
+                        "forwardingStyle", ForwardingStyle.ASYNCHRONOUS_REQUEST_TO_BROKER)
+                .build();
+
+        try (var tester = KroxyliciousTesters.mockKafkaKroxyliciousTester(
+                mockBootstrap -> singleRouteConfigWithVcFilter(mockBootstrap, filterDef), ROUTING_ENABLED);
+                var client = tester.simpleTestClient()) {
+
+            ApiVersionsResponseData apiVersions = new ApiVersionsResponseData();
+            apiVersions.apiKeys().add(new ApiVersionsResponseData.ApiVersion()
```
Comment (k-wall) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/RouteFilterCorrectnessIT.java]: Ditto - Fetch response mocking redundant.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/RouteFilterCorrectnessIT.java [hunk 16]
```diff
@@ -221,6 +236,345 @@ void routeFilterSendRequestWorksWithSingleRoute() {
...
+        var innerCounterDef = new NamedFilterDefinitionBuilder("c10-inner-counter", ResponseCountingFilterFactory.class.getName())
+                .withConfig("counterId", innerCounterId)
+                .build();
+        var config = nestedRouteWithResponseCounters(cluster.getBootstrapServers(), outerCounterDef, innerCounterDef);
+
+        try (var tester = KroxyliciousTesters.newBuilder(config).setFeatures(ROUTING_ENABLED).createDefaultKroxyliciousTester();
+                var producer = tester.producer(Map.of(DELIVERY_TIMEOUT_MS_CONFIG, 3_600_000))) {
+
+            // When
+            producer.send(new ProducerRecord<>(topic.name(), "key", "value")).get();
```
Comment (k-wall) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/RouteFilterCorrectnessIT.java:481]: nit: use assertJ's succeedsWithin.  We don't want a test suite that can hang forever.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/RouteFilterCorrectnessIT.java [hunk 17]
```diff
@@ -72,6 +78,14 @@
  * <p>C5: Route filters process only their own route's {@code InternalRequestFrame}s.
  * <p>C6: Same filter factory type on multiple routes does not cause cross-contamination.
  * <p>C7: Route filter {@code onResponse} is called for dynamically-dispatched responses.
+ * <p>C8: Sibling filters' {@code onResponse} observation of an out-of-band response is not
```
Comment (k-wall) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/RouteFilterCorrectnessIT.java:81]: I think the new ITs look thorough.

**Review by k-wall — APPROVED** [review]
Couple of suggestions, but looks like a huge improvement over what we had.  Thank you.
There's also a DCO issue.
